src/compas/robots/model/robot.py

```diff
@@ -575,6 +591,18 @@ def load_geometry(self, *resource_loaders, **kwargs):
 
                     if not shape.geometry:
                         raise Exception('Unable to load geometry for {}'.format(shape.filename))
+        self._geometry_loaded = True
```

Comment:
Using this flag isn't a good idea in a couple of cases.  First if the robot is defined purely using primitives, then `load_geometry` wouldn't need to be called.  Secondly, if a user adds a link manually (that is, not with `add_link`, but with creating a `Link` instance, etc) and gives its collision and visual elements a `filename` rather than a mesh (as unlikely as this is), this could lead to mysterious errors being raised.  So I think I need to essentially copy the nested for loops into `ensure_geometry`.

Reply:
Yes, agree with the inconvenience of the flag. The `ensure_*` methods are basically asserts, so, if there's the risk of stale state, then the nested loop suddenly feels like a good option ;)